ambry-store/src/main/java/com.github.ambry.store/BlobStoreCompactor.java

```diff
@@ -0,0 +1,899 @@
+    List<IndexEntry> validEntries = new ArrayList<>();
+    for (IndexEntry indexEntry : allIndexEntries) {
+      IndexValue value = indexEntry.getValue();
+      if (value.isFlagSet(IndexValue.Flags.Delete_Index)) {
+        // DELETE entry. Always valid.
+        validEntries.add(indexEntry);
+        // if this delete cannot be counted and there is a corresponding unexpired PUT entry in the same index segment,
+        // we will need to add it.
+        // NOTE: In PersistentIndex.markAsDeleted(), the expiry time of the put value is copied into the delete value.
+        // So it is safe to check for isExpired() on the delete value.
+        if (!deletesInEffect && !srcIndex.isExpired(value)) {
+          long putRecordOffset = value.getOriginalMessageOffset();
+          if (putRecordOffset != IndexValue.UNKNOWN_ORIGINAL_MESSAGE_OFFSET
+              && indexSegmentStartOffset.getOffset() <= putRecordOffset) {
+            BlobReadOptions options =
+                srcIndex.getBlobReadInfo(indexEntry.getKey(), EnumSet.allOf(StoreGetOptions.class));
+            Offset offset = new Offset(indexSegmentStartOffset.getName(), putRecordOffset);
+            IndexValue putValue = new IndexValue(options.getSize(), offset, options.getExpiresAtMs());
+            validEntries.add(new IndexEntry(indexEntry.getKey(), putValue));
+            options.close();
+          }
+        }
+      } else if (!srcIndex.isExpired(value)) {
+        // unexpired PUT entry.
+        if (deletesInEffect) {
+          FileSpan deleteSearchSpan =
+              new FileSpan(indexSegmentStartOffset, startOffsetOfLastIndexSegmentForDeleteCheck);
+          IndexValue searchedValue = srcIndex.findKey(indexEntry.getKey(), deleteSearchSpan);
+          if (!searchedValue.isFlagSet(IndexValue.Flags.Delete_Index)) {
+            // PUT entry that has not expired and is not considered deleted.
+            validEntries.add(indexEntry);
+          } else {
+            logger.trace("Skipping {} because it is a deleted PUT", indexEntry.getKey());
+          }
+        } else {
+          // valid PUT entry
+          validEntries.add(indexEntry);
+        }
+      } else {
+        logger.trace("Skipping {} because it is an expired PUT", indexEntry.getKey());
+      }
+    }
+    return validEntries;
+  }
+
+  /**
+   * Checks if a record already exists in the target log (swap spaces).
+   * @param key the {@link StoreKey} of the record.
+   * @param srcValue the {@link IndexValue} obtained from the application index.
+   * @return {@code true} if the record already exists in the target log, {@code false} otherwise.
+   * @throws StoreException if there is any problem with using the index.
+   */
+  private boolean alreadyExistsInTgt(StoreKey key, IndexValue srcValue) throws StoreException {
+    IndexValue tgtValue = tgtIndex.findKey(key);
+    return tgtValue != null && srcValue.isFlagSet(IndexValue.Flags.Delete_Index) == tgtValue.isFlagSet(
```

Comment:
I gave this some thought and the the situation you pointed out cannot occur for how the code is currently written (can verify offline) - so there aren't correctness issues. However, to make this function more general (so that code changes don't break it), I have made a modification.